diff -Naur fastnetmon-1.2.0.orig/src/fast_dpi.h fastnetmon-1.2.0/src/fast_dpi.h
--- fastnetmon-1.2.0.orig/src/fast_dpi.h	2022-03-29 23:00:24.000000000 +0200
+++ fastnetmon-1.2.0/src/fast_dpi.h	2022-04-19 10:38:59.612286571 +0200
@@ -1,7 +1,7 @@
 #ifndef FAST_DPI_H
 #define FAST_DPI_H
 
-#include "libndpi/ndpi_api.h"
+#include "ndpi_api.h"
 #include <stdlib.h>
 
 struct ndpi_detection_module_struct* init_ndpi();
