Description: Use pkg-config to find libndpi
Author: Benjamin Drung <benjamin.drung@profitbricks.com>
Bug: https://github.com/pavel-odintsov/fastnetmon/issues/658
Last-Update: 2017-05-08

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -233,15 +233,10 @@
 
     add_library(fast_dpi STATIC fast_dpi.cpp)
 
-    set(NDPI_INCLUDE_DIRS "/opt/ndpi/include/libndpi-1.7.1")
+    find_package(PkgConfig REQUIRED)
 
-    find_library(NDPI_LIBRARIES NAMES ndpi PATHS "/opt/ndpi/lib" NO_DEFAULT_PATH)
+    pkg_check_modules(NDPI REQUIRED libndpi)
 
-    if (NOT NDPI_LIBRARIES)
-        message(FATAL_ERROR "Could not find nDPI library")
-    endif()
-
-    link_directories("/opt/ndpi/lib")
     include_directories(${NDPI_INCLUDE_DIRS})
 
     add_definitions(-DENABLE_DPI)
